targets/TARGET_Analog_Devices/TARGET_ADUCM4X50/TARGET_ADUCM4050/api/trng_api.c

```diff
@@ -0,0 +1,120 @@
+ *     or more patent holders.  This license does not release you from the
+ *     requirement that you obtain separate licenses from these patent holders
+ *     to use this software.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES, INC. AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-
+ * INFRINGEMENT, TITLE, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL ANALOG DEVICES, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, PUNITIVE OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, DAMAGES ARISING OUT OF
+ * CLAIMS OF INTELLECTUAL PROPERTY RIGHTS INFRINGEMENT; PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+#if defined(DEVICE_TRNG)
+
+#include <stdlib.h>
+#include <adi_rng.h>
+#include <adi_pwr.h>
+#include "cmsis.h"
+#include "trng_api.h"
+
+// Sampling counter values
+// Prescaler: 0 - 10
+// LenReload: 0 - 4095
+#define TRNG_CNT_VAL	4095
+#define TRNG_PRESCALER 	2
+
+/* Data buffers for Random numbers */
+static uint32_t RngDevMem[(ADI_RNG_MEMORY_SIZE + 3)/4];
+
+void trng_init(trng_t *obj)
+{
+	ADI_RNG_HANDLE RNGhDevice;
+
+	// Open the device
+    adi_rng_Open(0,RngDevMem,sizeof(RngDevMem),&RNGhDevice);
+
+    // Set sample length for the H/W RN accumulator
+    adi_rng_SetSampleLen(RNGhDevice, TRNG_PRESCALER, TRNG_CNT_VAL);
+
+	// Disable buffering - single byte generation only
+	adi_rng_EnableBuffering(RNGhDevice, false);
+
+	// Enable the TRNG
+    adi_rng_Enable(RNGhDevice, true);
+
+	// Save device handle
+	obj->RNGhDevice = RNGhDevice;
+}
+
+void trng_free(trng_t *obj)
+{
+    ADI_RNG_HANDLE RNGhDevice = obj->RNGhDevice;
+
+    adi_rng_Enable(RNGhDevice, false);
+    adi_rng_Close(RNGhDevice);
+}
+
+int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
+{
+    ADI_RNG_HANDLE RNGhDevice = obj->RNGhDevice;
+    bool bRNGRdy;
+    uint32_t nRandomNum, i;
+    ADI_RNG_RESULT result;
+
+    for (i = 0; i < length; i++) {
+        // Loop until the device has data to be read
+        do {
+            result = adi_rng_GetRdyStatus(RNGhDevice, &bRNGRdy);
+            if (result != ADI_RNG_SUCCESS)
+            {
+                return -1;
+            }
+        } while (!bRNGRdy);
+
+        // Read the RNG
+        result = adi_rng_GetRngData(RNGhDevice, &nRandomNum);
+
+        if (result != ADI_RNG_SUCCESS)
+        {
+            return -1;
+        }
+
+        // Save the output
+		output[i] = (uint8_t)(nRandomNum & 0xFF);
+    }
+
```

Comment:
Please clear `nRandomNum` from the stack before returning, as otherwise the last random byte might leak.

Reply:
OK, will add the code to clear the variable.

Reply:
To avoid having the compiler get into our way and remove the zeroing, `nRandomNum` should be qualified `volatile`, or at least zeroed through accessing a `volatile`-qualified pointer (it can be outside the loop, though).

Reply:
The compiler wouldn't optimize this zeroing out as the address of nRandomNum is passed into the adi_rng_GetRngData function and the compiler cannot assume that the initialized value of nRandomNum isn't used inside this function. So I think it is safe.

Reply:
Please declare the variable as volatile. While it's neither very likely nor severe that the write to `nRandomNum` gets optimized out, it is possible (note that the compiler may assume `adi_rng_GetRngData` doesn't read it because it's uninitialized initially), and we should use best efforts.

Reply:
Fixed in the latest commit